Move gateway formatting to new function

martindekov opened the pull request:
Moving the gateway formatting to the new function
which means we no longer need to manually remove
`.openfaas` prefix if we are not on Kubernetes
now we just comment out `dnx_suffix` in
gateway_config.yml

Signed-off-by: Martin Dekov (VMware) <mdekov@vmware.com>

## Description

Revendored the SDK for `buildshiprun` `garbage-collect` `git-tar` `github-event` `github-push` `list-functions` and now gateway is made by getting `dns_suffix` and `gateway_url` then use the function to transform it in the right condition depending on existance of `dns_suffix`. Also added `dnsFormatter` for dashboard and moved configuration one level up so we have mainly `gateway_config.yml` for almost all of the configuration of the cloud

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Swarm:

Redeployed functions from above and committed change in repository and the building process was successful.

Removed installation from repository to see `garbage-collect` do its job

Saw output from `list-functions` it was as expected.

Also `github-status` successfully reports the status to github commit and checks API

Kubernetes: TO DO

Documentation: Updated

Javascript function: Added

## How are existing users impacted? What migration steps/scripts do we need?

If users want latest function they need to rebase their branches with master branch because the gateway_config.yml is changed so it wont work properly with newest changes.

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [x] read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] signed-off my commits with `git commit -s`
- [ ] added unit tests

alexellis commented:
Several rebase issues on this now.

martindekov commented:
Yeah I noticed will re-base and re-vendor again since the lock files are also bugged I think you should not touch Gopkg.lock but rather revendor

alexellis commented:
Let's come back to this after testing and deploying the new OF Cloud versions from today.

martindekov commented:
This really looks odd.. maybe I should rebase/reopen the PR and check whats up. This have been re-based couple of times already because it's been opened since couple of re-vendoring in previous merged PR.

alexellis commented:
FYI: I can see some "TODO" items in the description. If we don't get this ready within a few hours it'll get pushed into the code freeze until after 12th Oct.

martindekov commented:
Can't test this on kubernetes.. Need to ask someone from the community

alexellis commented on gateway_config.yml:
```diff
@@ -1,10 +1,12 @@
 environment:
 # URLs
-  gateway_url:  http://gateway.openfaas:8080/
+  gateway_url:  http://gateway:8080/
```
Please change these URLs so they match the original format.

```
gateway.openfaas
```

alexellis commented on garbage-collect/vendor/github.com/openfaas/openfaas-cloud/sdk/service.go:
```diff
@@ -8,3 +8,17 @@ import (
 	return fmt.Sprintf("%s-%s", strings.ToLower(owner), functionName)
 }
+
+func CreateServiceURL(URL, suffix string) string {
```
Let's add a test for backwards compatibility where people can call this method more than once without getting the suffix added twice.

```
http://gateway.openfaas:8080 = fn(http://gateway:8080, openfaas)
http://gateway.openfaas:8080 = fn(http://gateway. openfaas:8080, openfaas)
```

alexellis commented:
We need to consider the dashboard URLs also.

martindekov commented:
And for dashboard I see couple of overlaps which I am not sure what to do with since we showed that `../somefile.yml` works which are:

1. gateway_config.yml `basic_auth` == dashboard/stack.yml system-list-functions `basic_auth`
2. gateway_config.yml `gateway_public_url` == dashboard/stack.yml  system-dashboard `public_url`
3. gateway_config.yml `gateway_pretty_url` == dashboard/stack.yml  system-dashboard `pretty_url`
4. gateway_config.yml `secret_mount_path` == dashboard/stack.yml  system-dashboard `secret_mount_path`
I saw those overlap, I guess they are the same thing but still. I can amend the change in separate commit to the dashboard and replace dashboard variables to match the gatewa_config.yml files e.g. `public_url` -> `gateway_public_url` in dashboard code and so on

martindekov commented:
I amended the comment ^ with these two test cases found [service_test.go](https://github.com/openfaas/openfaas-cloud/blob/aa7056f8fbbc54645e81ebdd48e4fd111ccff83a/sdk/service_test.go) here:
```
		{
			title:       "URL formatted for Kubernetes",
			URL:         "http://gateway:8080",
			suffix:      "openfaas",
			expectedURL: "http://gateway.openfaas:8080",
		},
		{
			title:       "URL formatted for Kubernetes",
			URL:         "http://gateway.openfaas:8080",
			suffix:      "",
			expectedURL: "http://gateway.openfaas:8080",
		},
		{
			title:       "URL formatted for Kubernetes",
			URL:         "http://gateway.openfaas:8080",
			suffix:      "openfaas",
			expectedURL: "http://gateway.openfaas:8080",
		},
```
So this shows it is backward compatible. And about the dashboard once we settle on solution in the comment I left I can open another PR since I think this is already overcrowded(opinion)

alexellis commented:
I'm seeing merge conflicts on this now.

alexellis commented:
We appear to be getting a CI failure on this now.

martindekov commented:
Not related to the PR I will add those fields to the SDK and wait for release
```
function/handler.go:214:72: event.OwnerID undefined (type *sdk.Event has no field or method OwnerID)
function/handler.go:226:52: event.RepoURL undefined (type *sdk.Event has no field or method RepoURL)
function/handler.go:340:6: info.RepoURL undefined (type sdk.Event has no field or method RepoURL)
function/handler.go:343:7: info.OwnerID undefined (type sdk.Event has no field or method OwnerID)
```
Probably added manually to the sdk

martindekov commented:
This is my mistake

martindekov commented:
Hey there, I added tests with `jest` for the javascript function and additional function which checks if string variable is empty, added test cases to show expected behaviours for it too, in different PR this can be added to travis to be executed.

alexellis commented:
@martindekov this is a big change, I'd like to ask whether you feel it's important now that we've decided that OFC only runs on Kubernetes and not on Swarm? I feel like we created this PR in order to support Swarm.

OFC should be installable and functioning on an alternative namespace other than `.openfaas`, is it that way today? cc @Waterdrips

alexellis commented:
Feel free to re-open

martindekov commented:
Nah I believe now Kubernetes is the thing and this adds unnecessary configuration which might make things more complex.
